src/zmalloc.c

```diff
@@ -88,10 +89,34 @@ void zlibc_free(void *ptr) {
 #define dallocx(ptr, flags) je_dallocx(ptr, flags)
 #endif
 
-#define update_zmalloc_stat_alloc(__n) atomic_fetch_add_explicit(&used_memory, (__n), memory_order_relaxed)
-#define update_zmalloc_stat_free(__n) atomic_fetch_sub_explicit(&used_memory, (__n), memory_order_relaxed)
+#if __STDC_NO_THREADS__
+#define thread_local __thread
+#else
+#include <threads.h>
+#endif
+
+/* A thread-local storage which keep the current thread's index in the used_memory_thread array. */
+static thread_local int thread_index = -1;
+static size_t used_memory_thread[MAX_THREADS_NUM];
```

Comment:
in case allocation and free are done in different threads. how will use of unsigned work in case we might get negative values and wraparound?

Reply:
we sum all used_memory_thread, the aggregated value should be fine.

Reply:
but negative values can be placed into unsigned size_t? maybe you should use ssize_t?

Reply:
It should have no impact the final sum value, the thread variable just an internal state. The total sum value should be positive. But if you want, I can change them to signed number.
```c
size_t a = -1;
size_t b = 1;
size_t c = a + b;
assert(c == 0);
```

Reply:
I believe unsigned is OK, because the sum of these is always positive. Only the sum should be observable.